### ISSUE TITLE
[inductor] Wrong argmax after fused trailing-dim reduction on non-contiguous inputs (incl. non-unit-stride reduction window) — generic reduction fusion ignores input strides, CPU and CUDA

### ISSUE THREAD

**Issue by BUAA-Wander** [pytorch/pytorch#193661]
### 🐛 Describe the bug

**Relationship to #193492**: this is the same *mechanism* (the fused kernel assumes a contiguous interpretation of the input and ignores the real strides) but a **different lowering** — the generic reduction fusion (`mean`/`var`/`sum`/`logsumexp`/`amax` + `argmax`), not the `_adaptive_avg_pool2d` path. The contiguous guard proposed in #193492 is added at the `_adaptive_avg_pool2d` lowering and does not apply to these graphs; the reduction path needs its own fix/regression coverage. I asked in the #193492 thread whether this path should be tracked separately ([[comment](https://github.com/pytorch/pytorch/issues/193492#issuecomment-5299760647)](https://github.com/pytorch/pytorch/issues/193492#issuecomment-5299760647)); opening this issue here so it doesn't get lost in the pool discussion. Happy to close this and fold everything back into #193492 if maintainers prefer a single tracking issue.

`torch.compile` with the Inductor backend returns a wrong `argmax` index when a trailing-dim reduction (`mean`, `var`, `sum`, `logsumexp`, also `amax`) is fused with `argmax` and the reduction input is a **non-contiguous view**. Eager returns the correct index; the compiled index points at a value that is **not** the extremum of the reduction output. The unfused reduction matches eager exactly (sanity `maxdiff ≈ 0`), so the mis-indexing is introduced by the fused kernel, whose generated input indexing assumes a contiguous layout and ignores the real strides. The trigger is deterministic (fixed seed, no randomness).

The previously-reported family instances used non-contiguous leading dims with a **stride-1** reduction window (e.g. `transpose`). The new instance below additionally has a **non-unit-stride reduction window** (`[..., ::2]`, stride 2 on the reduced dim) combined with non-contiguous leading dims. Neither axis triggers on its own: the same `[..., ::2]` slice on a contiguous input passes, and a fully contiguous input passes. As a scope hint for the regression test, `expand`/stride-0 views are **not** affected (all pass).

### To reproduce

```python
import torch

def fn(x):
    return torch.argmax(torch.mean(x, dim=-1))

torch.manual_seed(123)
x = torch.rand([2, 4, 4, 8]).transpose(0, 1).contiguous().transpose(0, 1)[..., ::2]
# shape (2, 4, 4, 4), strides (32, 64, 8, 2)
# non-contiguous leading dims AND a stride-2 reduction window

print("eager   =", fn(x).item())                # 12 (correct)
print("compiled=", torch.compile(fn)(x).item()) # 24 (wrong)
```

Truth check: value at the returned index vs. the true maximum of `mean(x, dim=-1)` (eager index points at the maximum; compiled index does not):

```text
eager    index 12 -> value 0.828343 (== max)
compiled index 24 -> value 0.596202 (!= max)
```

Scope / additional observations (all fixed seed 123, caches disabled, unfused reductions matching eager exactly):

- The same input mismatches for `var`, `sum`, `logsumexp` (dim=-1) and for `mean`, `sum`, `var` (dim=[-1, -2]): **7/7 reduction variants**, on CPU and CUDA, torch 2.13.0 and 2.5.1 (CPU indices identical across versions, e.g. mean: eager 12 / compiled 24 / true 12; CUDA also reproduces).
- Controls: the same `[..., ::2]` slice on a contiguous input passes, a fully contiguous input passes, and `expand`/stride-0 views pass — the trigger is the combination of non-contiguous leading dims and a non-unit-stride reduction window.
- The earlier family instances (non-contiguous leading dims, stride-1 reduction window) are also affected; see the #193492 thread for the family matrix.

Kernel evidence (CPU, torch 2.13.0): the aten graph contains only `mean.dim -> argmax` (the transpose/slice views are metadata and never enter the graph), while the fused kernel `cpp_fused_argmax_mean_0` reads the reduction window with a flat, contiguous interpretation:

```cpp
for (int64_t x0 = 0; x0 < 32; x0++) {   // x0 = flattened 2*4*4 leading positions
    auto tmp0 = in_ptr0[8 * x0];        // mean over in_ptr0[8*x0 + {0,2,4,6}]
    auto tmp1 = in_ptr0[2 + 8 * x0];
    auto tmp3 = in_ptr0[4 + 8 * x0];
    auto tmp5 = in_ptr0[6 + 8 * x0];
    ...
    tmp_acc0 = argmax_combine(tmp_acc0, mean, x0);
}
```

The true offset of element (i, j, k, k′) with strides `(32, 64, 8, 2)` is `32i + 64j + 8k + 2k′`, which the fused flat reads do not match. The wrong reads stay in bounds (max flat offset 254 < storage size 256), so the result is silently wrong — no OOB, no exception.

### Error logs

No exception — silent wrong result (wrong index returned).

### Versions

```text
PyTorch version: 2.13.0+cu130 and 2.5.1+cu121 (both reproduce, CPU and CUDA)
OS: Linux (AutoDL container)
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by SachaKefif** [comment]
Hi, I can reproduce this issue on CPU with the provided example. I’d be interested in working on a fix and adding regression coverage if this issue is open for contribution. Would it be okay for me to take this?

**Comment by bobrenjc93** [comment]
@SachaKefif Thanks for confirming the CPU reproduction. Yes, please go ahead with a fix and regression coverage, and reference #193661 in the PR. Device-generic coverage for the shared reduction-fusion path would be ideal if practical. This remains in triage review because it is a silent-correctness issue.

(Follow-up posted by Codex.)

[Human note] Just to be clear, while I’m supportive of this, I’m not volunteering to review these PRs. Codex drafted these comments during triage, and they seemed reasonable at the time. Looking back, though, I realize they could easily be interpreted as me signing up to review them. I’ll avoid leaving comments like this going forward. Sorry for the confusion!

**Comment by SachaKefif** [comment]
Hi ! I already linked the PR #193680

**Comment by eellison** [comment]
Let's make sure the fix to this also covers https://github.com/pytorch/pytorch/issues/193492.

cc @jananisriram who is assigned, @karthickai who did first version of this. can you shepherd along https://github.com/pytorch/pytorch/pull/193680 

Resolution after looking into the issue:

  Root cause is a time-of-check bug:

  1. At lowering, [`_make_reduction_inner`](.../lowering.py#L7279-L7304)
     decides the argmax doesn't need an explicit logical index (#165983's
     guard) — the input is an unrealized fused producer, nothing looks
     non-contiguous. The index stays implicit: the row-major flatten of the
     reduction loops.
  2. Later, the scheduler's [`pick_loop_order`](.../scheduler.py#L4199)
     (via `ComputedBuffer.simplify_and_reorder`) reorders the reduction loops to
     match the input's physical strides. The loads are rewritten correctly, but
     the implicit index just re-flattens the new loop order — wrong index, right
     value. 

  Suggested fix:

  1. Always emit `argreduce(value, logical_index)` and delete the guard — the
     index becomes a real expression that loop transforms rewrite like any
     other. #193680 does this and fixes both issues.
  2. Make the reduction cache correct: the [Triton reduction CSE key](.../triton.py#L5559)
     omits the index, so `x.argmax()` / `x.t().argmax()` (same value, different
     index mapping) can merge and return wrong results. The key must include
     the index, as [CPP already does](.../cpp.py#L2484-L2488),
     with a regression test. Blocker for #193680.
  3. Later optimization to remove the overhead: a `LoopBody` rewrite right
     before kernel codegen (after all loop transforms) — if the carried index
     is provably the row-major flatten of the body's reduction vars, drop it
     and use the native counter. Proof failure costs index arithmetic, never
     correctness.

**Comment by sylvesterkaczmarek** [comment]
@eellison I took the remaining Triton CSE blocker in #194093. The PR now includes the logical index in Triton's reduction cache key, matching the C++ codegen, and adds a fused regression for x.argmax() together with x.t().argmax() so the same value expression cannot be reused across different logical index mappings.
